WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorViewModel.swift

```diff
@@ -1,24 +1,164 @@
+         stores: StoresManager = ServiceLocator.stores,
+         storageManager: StorageManagerType = ServiceLocator.storageManager) {
+        self.siteID = siteID
         self.wpAdminTaxSettingsURLProvider = wpAdminTaxSettingsURLProvider
+        self.stores = stores
+        self.storageManager = storageManager
+        self.paginationTracker = PaginationTracker(pageFirstIndex: 1, pageSize: 25)
+
+        configureResultsController()
+        configurePaginationTracker()
+        configureFirstPageLoad()
     }
 
     /// WPAdmin URL to navigate user to edit the tax settings
     var wpAdminTaxSettingsURL: URL? {
         wpAdminTaxSettingsURLProvider.provideWpAdminTaxSettingsURL()
     }
+
+    private lazy var resultsController: ResultsController<StorageTaxRate> = {
+        let predicate = NSPredicate(format: "siteID == %lld", siteID)
+        let sortDescriptorByID = NSSortDescriptor(keyPath: \StorageTaxRate.id, ascending: true)
+        let resultsController = ResultsController<StorageTaxRate>(storageManager: storageManager,
+                                                                    matching: predicate,
+                                                                    sortedBy: [ sortDescriptorByID])
+        return resultsController
+    }()
+
+    func onLoadNextPageAction() {
+        paginationTracker.ensureNextPageIsSynced()
+    }
+
+    func onRefreshAction(completion: @escaping () -> Void) {
+        paginationTracker.resync(reason: nil) {
+            completion()
+        }
+    }
+}
+
+extension NewTaxRateSelectorViewModel: PaginationTrackerDelegate {
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?) {
+        transitionToSyncingState()
+
+        let action = TaxAction.retrieveTaxRates(siteID: siteID, pageNumber: pageNumber, pageSize: pageSize) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let results):
+                let hasNextPage = results.count == pageSize
+                onCompletion?(.success(hasNextPage))
+
+                if self.taxRateViewModels.isEmpty {
+                    self.syncState = .empty
+                } else if results.isEmpty {
+                    // We had results previously, but we didn't have any on this page request. Transition to results to stop the syncing visuals.
+                    self.transitionToResultsUpdatedState()
+                }
+            case .failure(let error):
+                DDLogError("⛔️ Error synchronizing tax rates: \(error)")
+                onCompletion?(.failure(error))
+            }
+        }
+        stores.dispatch(action)
+    }
+}
+
+private extension NewTaxRateSelectorViewModel {
+    func configurePaginationTracker() {
+        paginationTracker.delegate = self
+    }
+
+    func configureFirstPageLoad() {
+        // Listens only to the first emitted event.
+        onLoadTrigger.first()
```

Comment:
So this constrains the `onAppear` to act like `onLoad`. Perhaps you could move the `.first()` to the definition of the `onLoadTrigger` property instead? That might remove the mental overhead while reading it here...

The comment could then be moved to the `onAppear` block in the view, to explain why we're linking appear to load.

Reply:
That makes sense, done in cd70c4ddda9998ca87e2d9af1ae724bf814e876d